### PR TITLE
Update CI to test against current Python/Django/DRF

### PR DESCRIPTION
**Pull request by philipstarkey**
With the release of [DRF 3.15](https://www.django-rest-framework.org/community/release-notes/#315x-series), bringing official support for Django 4.2+, I thought it would be a good opportunity to update the CI config of this project to test against support versions of Python/Django/DRF and drop official support for versions of those dependencies that are no longer getting updates.

## Changes
- Updated CI to run on currently supported Python versions (3.8+)
- Drop support for unsupported Python versions
- Updated CI to test against currently supported Django (3.2, 4.2, 5.0) and Django Rest Framework versions.
- Drop support for unsupported Django versions
- Updated CI dependencies to latest versions (required for Python 3.12 support)
- Updated dev dependencies to latest versions (required for Python 3.12 support)
- Updated documentation to reflect the requirements changes
- Updated PyPI metadata to reflect updated requirements
- Fix bug where `black` return status wasn't causing CI to fail
- Minor formatting changes due to upgrade of `black` version
- Update of `re_path` import in test
- Bumped version to 3.0.0 (major release due to dropping support for older versions)

### CODE REVIEW

**Comment by evenicoulddoit** [comment]
Hi @philipstarkey , thanks so much for this. Changes look great. I'm guessing the stylistic ones are just to keep Black happy. I'm using Ruff professionally now, it occurs to me that I could upgrade to this, plus go the `pyproject.toml` route

**Review by evenicoulddoit — COMMENTED** [review]
Just one question about `pytz` inclusion. Also, are you able to update the CHANGELOG? I couldn't remember, but I think I've been handling version updates manually, but always as part of a separate commit. So, I think we should remove the `__version__` being changed here :)

File: requirements-test.txt [hunk 1]
```diff
@@ -2,8 +2,10 @@
...
-djangorestframework>=3.9.4
-pytest-cov==2.6.1
-pytest-django==3.4.8
-pytest==4.4.1
+Django>=3.2,<5.1
+djangorestframework>=3.12
+pytest-cov==4.1.0
+pytest-django==4.8.0
+pytest==8.0.0
+pytz==2024.1
```
[the remark this answers is not in the document]
Comment (philipstarkey) [reply requirements-test.txt]: That is a very good question! I was sure it was needed to stop a test failing as something was relying on `pytz` being installed by another dependency....Ah it looks like `rest_framework` has done some things related to `pytz` support. Maybe it was only an issue with a combination of DRF/Django that we're not testing anymore.

Regardless of the reason, I've removed it again and everything still seems to pass given what I have configured now.

[the remark this answers is not in the document]
Comment (evenicoulddoit) [reply requirements-test.txt]: Great, thanks! I'm also aware that we _should_ be able to handle things with Python builtins now (Zoneinfo etc.)

**Comment by philipstarkey** [comment]
> I'm guessing the stylistic ones are just to keep Black happy. I'm using Ruff professionally now, it occurs to me that I could upgrade to this, plus go the `pyproject.toml` route

Yes! The stylistic change were to keep `black` happy. I'm not sure if that was due to the `black` version change, or just because `black` wasn't being passed `--check` so was never raising an exit code that would cause `tox` to fail. I'm also using Ruff professionally now - it does seem like it's the future.

> Also, are you able to update the CHANGELOG? I couldn't remember, but I think I've been handling version updates manually, but always as part of a separate commit. So, I think we should remove the `__version__` being changed here :)

I have updated the change log and reverted `__version__`. Let me know if there is anything else I can do!

**Pull request merged**